torch/csrc/autograd/engine.cpp

```diff
@@ -227,6 +228,12 @@ auto Engine::thread_main(GraphTask *graph_task) -> void {
   // Why the test on graph_task->outstanding_tasks?  See
   // Note [Reentrant backwards]
   while (!graph_task || graph_task->outstanding_tasks > 0) {
+
+    // NO_DEVICE worker will wait for tasks generated by other device while graph_task is not done
+    if (worker_device == NO_DEVICE && queue->heap.empty()) {
```

Comment:
This is unsafe. Access to `queue->heap` should be protected by the mutex. Right now you can prematurely break because some task might take a job from the queue and be in the middle of processing it, you'll see that it's empty and break out, while the other thread will finish processing and push more tasks into the queue.